bitwarden_access_manager.py

```diff
@@ -0,0 +1,634 @@
+#!/usr/bin/env python3
+
+"""
+Bitwarden Access Manager
+
+CONFIGURATION
+
+[[member]]
+member_id = "2564c11f-fc1b-4ec7-aa0b-afaf00a9e4a4"
+member_name = "yan"
+email = "yan.68@hotmail.fr"
+type = 2 # member
+access_all = false
+groups = ["group1", "group2"]
+
+[[member]]
+member_id = "856cba2d-cae1-40e7-96cc-afaf00a8a4cb"
+member_name = "yunkel"
+email = "yunkel68@hotmail.fr"
+type = 0 # owner
+access_all = true
+groups = ["group1"]
+
+[[group]]
+group_id = "c6a13b93-edc1-4c3b-9fc5-afaf00a8d33f"
+group_name = "group1"
+access_all = false
+
+[[group]]
+group_id = "39b48ab2-81fd-40eb-87e9-afb0000110f3"
+group_name = "group2"
+access_all = false
+
+[[collection]]
+collection_id = "50351c20-55b4-4ee8-bbe0-afaf00a8f25d"
+external_id = "collection1"
+member_access = [
+  { member_id = "2564c11f-fc1b-4ec7-aa0b-afaf00a9e4a4", member_name = "yan", group = "c6a13b93-edc1-4c3b-9fc5-afaf00a8d33f" },
```

Comment:
This is not about direct member-to-collection access. If allow a member to access a collection directly, the member type will be changed to custom. We don't permit members with custom type, therefore we do not need to check member direct access to collection.

 Bitwarden doesn't tell you all users who have access to a collection because they are in group. I added [get_collection_members](https://github.com/ChorusOne/github-access-manager/pull/10/files#diff-7bfcbcb708fa5299e4545d8f8335bc2ab16d37b5b7e70edd0262e41d040b133aR310) method to do it.

I though it would be useful to get this information closer to the list of group accessing the collection. This way you don't need to scroll the config, find each users accessing the collection and check in which group they are. But you're right we don't need to list it in the config.

Reply:
> This is not about direct member-to-collection access. If allow a member to access a collection directly, the member type will be changed to custom. We don't permit members with custom type, therefore we do not need to check member direct access to collection.

I double check this statement and it was wrong. The member type doesn't change to custom when we  give access to a member to a collection.

Currently `member_access` only list members that are part of a group accessing a collection. It doesn't include users with a direct access to a collection.

We should add users that have direct access to a collection and write in the config their permissions to the collection.

We could create another list `member_direct_access` which is optional to have a clear distinction with user accessing the collection because they are part of a group.
```
[[collection]]
collection_id = "30e5cf56-9a4f-4332-8535-afb70077dbbb"
external_id = "collection2"
member_group_access = [
  { member_name = "yunkel"},
]
member_direct_access = [
  { member_name="yan", access="write"}
]
group_access = [
  { group_name = "group1", access = "readonly" },
]
```

We could also use the existing `member_access` list, keep the `group` key for members part of a group and add a `access` key to members with direct access to the collection.
```
[[collection]]
collection_id = "30e5cf56-9a4f-4332-8535-afb70077dbbb"
external_id = "collection2"
member_access = [
  { member_name = "yunkel", group="group1" },
  { member_name = "yan", access="write" }
]
group_access = [
  { group_name = "group1", access = "readonly" },
]
```

I prefer the first option. @ruuda what do you tink ?

Reply:
> Bitwarden doesn't tell you all users who have access to a collection because they are in group.

That is fine, we are only interested in the groups that have access to a collection (as long as no individuals have _direct_ access).

> I prefer the first option. @ruuda what do you think ?

Let me propose a third option :P

```toml
[[group]]
name = "group1"
group_id = "..."

[[user]]
name = "yunkel"
# I’m still not sure about this this, maybe groups should instead list their members instead of users listing their groups.
# Maybe if there’s a group "admins" you want to see at a glance who they are ... but then you can also just grep for it.
# Anyway that was not the point of this comment.
groups = ["group1"]

[[collection]]
collection_id = "30e5cf56-9a4f-4332-8535-afb70077dbbb"
external_id = "collection2"
# This lists only _direct_ access, i.e. not members that have access through being member of a group.
member_access = [
  { member_name = "yan", access="write" }, # Probably need to have the user id in there too
]
group_access = [
  { group_name = "group1", access = "readonly" },
]
```
This is the same as what we do for GitHub repository access, it lists the groups, and members that have direct access, but not members that have access through a group.

If you want to see at a glance per collection which users have access to it, i.e. expand the groups, right now I don’t really feel the need for that, but if we need it, I would add a new subcommand that can print this. Then we can keep the config file free of redundant information.

Reply:
This https://github.com/ChorusOne/github-access-manager/pull/10/commits/afdb1ca932a1425667311a77d5737c2a34da75c6 introduce listing of direct member access to a collection. I omitted the listing of members with the role `ADMIN` or `OWNER` because they have [access to all collections by default](https://bitwarden.com/help/user-types-access-control/#member-roles).